Use finegrain allocator by default

abouteiller opened the pull request:
This revamps the hip-allocator selection logic, and set the hipAllocatorFineGrained as the default, we can set the Uncached mode with a ccmake debug control.

abouteiller committed:
Use FineGrained allocator for heap by default, consolidate all types of
allocators under saner cmake controls

Co-authored-by: Yiltan <ytemucin@amd.com>

abouteiller commented on cmake/rocshmem_config.h.in:
```diff
@@ -29,12 +29,21 @@
 #cmakedefine USE_HOST_SIDE_HDP_FLUSH
+
+#define HEAP_IS_COHERENT
+#if defined(USE_HEAP_HOST) || defined(USE_HEAP_HOST_NONCOHERENT) || defined(USE_HEAP_DEVICE_COARSEGRAIN)
```
Please review this carefully :)

Yiltan commented on cmake/rocshmem_config.h.in:
```diff
@@ -29,12 +29,21 @@
+
+#define HEAP_IS_COHERENT
+#if defined(USE_HEAP_HOST) || defined(USE_HEAP_HOST_NONCOHERENT) || defined(USE_HEAP_DEVICE_COARSEGRAIN)
+/* refer to for why https://rocm.blogs.amd.com/software-tools-optimization/mi200-memory-space/README.htmla
```
```suggestion
/* refer to for why https://rocm.blogs.amd.com/software-tools-optimization/mi200-memory-space/README.html
```

abouteiller committed:
Uncached may not be only for debug

Need to include the rocshmem config otherwise produce an inconsistent
build with different allocators used in different files

abouteiller committed:
Undo this pr adding presumably useless hip_host_allocator_noncoherent

abouteiller committed:
Rename HEAP_IS_COHERENT/USE_COHERENT_HEAP to USE_HDP_FLUSH as the former
was misleading

abouteiller committed:
Remove unused __roc_inv()
